modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java

```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
+        if (ctx.channel().isWritable()) {
+            doFlush(ctx);
+        }
+        ctx.fireChannelWritabilityChanged();
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        Channel channel = ctx.channel();
+        if (channel.isWritable() || channel.isActive() == false) {
+            doFlush(ctx);
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        doFlush(ctx);
+        super.channelInactive(ctx);
+    }
+
+    private void doFlush(ChannelHandlerContext ctx) {
+        final Channel channel = ctx.channel();
+        if (channel.isActive() == false) {
+            if (currentWrite != null) {
+                currentWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            WriteOperation queuedWrite;
+            while ((queuedWrite = queuedWrites.poll()) != null) {
+                queuedWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            return;
+        }
+        while (channel.isWritable()) {
+            if (currentWrite == null) {
+                currentWrite = queuedWrites.poll();
+            }
+            if (currentWrite == null) {
+                break;
+            }
+            final WriteOperation write = currentWrite;
+            final ByteBuf writeBuffer;
+            if (write.buf.readableBytes() == 0) {
+                write.promise.trySuccess();
+                currentWrite = null;
+                continue;
+            }
+            final int bufferSize = Math.min(write.buf.readableBytes(), 1 << 18);
+            writeBuffer = ctx.alloc().directBuffer(bufferSize, bufferSize);
+            writeBuffer.writeBytes(write.buf, bufferSize);
+            writeBuffer.retain();
+            final ChannelFuture writeFuture;
+            try {
+                writeFuture = ctx.write(writeBuffer);
+            } finally {
+                writeBuffer.release();
+            }
+            if (write.buf.readableBytes() == 0) {
+                currentWrite = null;
+                writeFuture.addListener((ChannelFutureListener) future -> {
+                    if (future.isSuccess()) {
+                        write.promise.trySuccess();
+                        doFlush(ctx);
+                    } else {
+                        write.promise.tryFailure(future.cause());
+                    }
+                });
+            } else {
+                writeFuture.addListener(f -> {
+                    if (f.isSuccess() == false) {
+                        write.promise.tryFailure(f.cause());
+                    } else {
+                        doFlush(ctx);
+                    }
+                });
+            }
+            ctx.flush();
+            if (channel.isActive() == false) {
+                WriteOperation queuedWrite;
+                while ((queuedWrite = queuedWrites.poll()) != null) {
```

Comment:
I think we should also notify currentWrite if it is not null and thus this is the same as is done in the beginning of this method (extract to method?).

Reply:
`currentWrite` will have its callback fail on the event loop because we just enqueued a flush for it. If we touch it here as well it should be a noop.